cpp/src/arrow/engine/catalog.h

```diff
@@ -0,0 +1,94 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#pragma once
+
+#include <memory>
+#include <string>
+#include <unordered_map>
+#include <utility>
+#include <vector>
+
+#include "arrow/type_fwd.h"
+#include "arrow/util/variant.h"
+
+namespace arrow {
+
+namespace dataset {
+class Dataset;
+}
+
+namespace engine {
+
+/// Catalog is made of named Table/Dataset to be referenced in LogicalPlans.
+class Catalog {
+ public:
+  class Entry;
+
+  static Result<std::shared_ptr<Catalog>> Make(const std::vector<Entry>& tables);
+
+  Result<Entry> Get(const std::string& name) const;
+  Result<std::shared_ptr<Schema>> GetSchema(const std::string& name) const;
+
+  class Entry {
+   public:
+    enum Kind {
```

Comment:
Is it actually necessary to accept both? Can't you just create an in-memory Dataset from a Table? Sounds like it would be potentially useful in any case.

Reply:
The InMemoryDataset couldn't take a table when I created the branch. I rebased with ARROW-7886 and now can.

Reply:
I remember why I made it like this; if we want the compute module to work without the dataset component (shared library), we need this.

Reply:
I think Dataset should eventually land in the top namespace, and we should build it always.

Reply:
That sounds fine, but in that case the parquet dependency of `arrow::dataset::` must be made optional (this shouldn't be hard; I think it just needs an `#ifdef` in arrow/dataset/api.h)